picoquic/frames.c

```diff
@@ -914,75 +914,93 @@ uint8_t* picoquic_decode_stream_frame(picoquic_cnx_t* cnx, uint8_t* bytes, const
 
 picoquic_stream_head* picoquic_find_ready_stream(picoquic_cnx_t* cnx)
 {
-    picoquic_stream_head* stream = NULL;
+    picoquic_stream_head* start_stream = cnx->first_stream;
+    picoquic_stream_head* found_stream = NULL;
 
-    for (int nb_pass = 0; nb_pass < 2; nb_pass++) {
-        stream = cnx->first_stream;
 
-        if (nb_pass == 0) {
-            /* Skip to the first non visited stream */
-            while (stream && stream->stream_id <= cnx->last_visited_stream_id) {
-                stream = stream->next_stream;
-            }
-        }
+    /* Skip to the first non visited stream */
+    while (start_stream && start_stream->stream_id <= cnx->last_visited_stream_id) {
+        start_stream = start_stream->next_stream;
+    }
+
+    if (start_stream == NULL) {
+        start_stream = cnx->first_stream;
+    }
 
-        while (stream) {
-            if ((cnx->maxdata_remote > cnx->data_sent && stream->sent_offset < stream->maxdata_remote &&
-                (stream->is_active ||
-                (stream->send_queue != NULL && stream->send_queue->length > stream->send_queue->offset) ||
-                (stream->fin_requested && !stream->fin_sent))) ||
-                    (stream->reset_requested && !stream->reset_sent) ||
+    /* Look for a ready stream */
+    if (start_stream != NULL) {
+        picoquic_stream_head* stream = start_stream;
+        do {
+            if ((cnx->maxdata_remote > cnx->data_sent && stream->sent_offset < stream->maxdata_remote && (stream->is_active || 
+                    (stream->send_queue != NULL && stream->send_queue->length > stream->send_queue->offset) ||
+                    (stream->fin_requested && !stream->fin_sent))) ||
+                (stream->reset_requested && !stream->reset_sent) ||
                 (stream->stop_sending_requested && !stream->stop_sending_sent)) {
                 /* Something can be sent */
                 /* if the stream is not active yet, verify that it fits under
                  * the max stream id limit */
                  /* Check parity */
-                if (IS_CLIENT_STREAM_ID(stream->stream_id) == cnx->client_mode) {
-                    if (stream->stream_id <= cnx->max_stream_id_bidir_remote) {
-                        break;
-                    }
-                }
-                else {
+                if (IS_CLIENT_STREAM_ID(stream->stream_id) != cnx->client_mode ||
+                    (IS_BIDIR_STREAM_ID(stream->stream_id) && stream->stream_id <= cnx->max_stream_id_bidir_remote) ||
+                    (!IS_BIDIR_STREAM_ID(stream->stream_id) && stream->stream_id <= cnx->max_stream_id_unidir_remote)) {
```

Comment:
You will probably hate my style, but just for the record, I would had written this as:
```
stream->stream_id <= (IS_BIDIR_STREAM_ID(stream->stream_id) ? cnx->max_stream_id_bidir_remote : cnx->max_stream_id_unidir_remote)
```

Reply:
Yes, I could save a couple lines. I assume that the compiler will optimize all that, so i went for the easy-to-read form, but having a single line is not that bad.